script/util/makeAlpRegex.sh

```diff
@@ -0,0 +1,16 @@
+#!/bin/bash
+
+FILE=$1
+
+if [ ! -e $FILE ]; then
+  echo "$FILE not found"
+  exit 1
+fi
+
+REGEX=""
+
+for e in `cat $FILE | egrep 'e\.(GET|POST|PATCH|PUT|DELETE)\("/.*"' | sed 's/^.*"\//\//g' | sed 's/".*$//g' | sed 's/:[A-Za-z0-9_]*/\[\^\/]\+/g' | sort -r | uniq`; do
```

Comment:
必ずしも`e.GET/POST`の形式ではないね👀
`.GET/POST`の形式に絞るのはどうだろう？

[memo] isucon12f/.../main.go
https://github.com/isucon/isucon12-final/blob/main/webapp/go/main.go

I tried, but...
[result]
<img width="742" alt="Screenshot 2023-11-27 at 8 55 00" src="https://github.com/sugar-cat7/isucon-tool/assets/27912121/d886919e-8af3-4976-aa12-69ba7dd5e40a">

Reply:
[IMO] 下記のようにするのはどうだろう！

```suggestion
for e in `cat $FILE | grep -E '[a-zA-Z0-9_]+\.(GET|POST|PATCH|PUT|DELETE)\("/.*"' | sed 's/^.*"\//\//g' | sed 's/".*$//g' | sed 's/:[A-Za-z0-9_]*/\[\^\/]\+/g' | sort -r | uniq`; do
```

<img width="784" alt="Screenshot 2023-11-27 at 9 06 51" src="https://github.com/sugar-cat7/isucon-tool/assets/27912121/81196925-6700-466c-a320-1a4533b0bcd7">

Reply:
isucon13-finalでの確認。スクリプト一見問題ない✅
```bash
❯❯ ./makeAlpRegex.sh ~/Desktop/main.go
/api/user/me,/api/user/[^/]+/theme,/api/user/[^/]+/statistics,/api/user/[^/]+/livestream,/api/user/[^/]+/icon,/api/user/[^/]+,/api/tag,/api/register,/api/payment,/api/login,/api/livestream/search,/api/livestream/reservation,/api/livestream/[^/]+/statistics,/api/livestream/[^/]+/report,/api/livestream/[^/]+/reaction,/api/livestream/[^/]+/ngwords,/api/livestream/[^/]+/moderate,/api/livestream/[^/]+/livecomment/[^/]+/report,/api/livestream/[^/]+/livecomment,/api/livestream/[^/]+/exit,/api/livestream/[^/]+/enter,/api/livestream/[^/]+,/api/livestream,/api/initialize,/api/icon
```

isucon13-final中に手動で作成したもの
```bash
/api/user/[^/]+/statistics,/api/livestream/\d+/livecomment,/api/livestream/\d+/reaction,/api/user/[^/]+/icon,/api/livestream/\d+/enter,/api/livestream/\d+/exit,/api/livestream/\d+/report,/api/livestream/\d+/ngwords,/api/livestream/\d+/moderate,/api/user/[^/]+/theme
```

違いをGPT4に解説してもらった

==============================
### リストA（makeAlpRegex.sh）
1. `/api/user/[^/]+/statistics`
2. `/api/livestream/\d+/livecomment`
3. `/api/livestream/\d+/reaction`
4. `/api/user/[^/]+/icon`
5. `/api/livestream/\d+/enter`
6. `/api/livestream/\d+/exit`
7. `/api/livestream/\d+/report`
8. `/api/livestream/\d+/ngwords`
9. `/api/livestream/\d+/moderate`
10. `/api/user/[^/]+/theme`

### リストB（手動で作成したやつ）
1. `/api/user/me`
2. `/api/user/[^/]+/theme`
3. `/api/user/[^/]+/statistics`
4. `/api/user/[^/]+/livestream`
5. `/api/user/[^/]+/icon`
6. `/api/user/[^/]+`
7. `/api/tag`
8. `/api/register`
9. `/api/payment`
10. `/api/login`
11. `/api/livestream/search`
12. `/api/livestream/reservation`
13. `/api/livestream/[^/]+/statistics`
14. `/api/livestream/[^/]+/report`
15. `/api/livestream/[^/]+/reaction`
16. `/api/livestream/[^/]+/ngwords`
17. `/api/livestream/[^/]+/moderate`
18. `/api/livestream/[^/]+/livecomment/[^/]+/report`
19. `/api/livestream/[^/]+/livecomment`
20. `/api/livestream/[^/]+/exit`
21. `/api/livestream/[^/]+/enter`
22. `/api/livestream/[^/]+`
23. `/api/livestream`
24. `/api/initialize`
25. `/api/icon`

### 違い
- リストAには、リストBにはないパターンが含まれています。例えば `/api/livestream/\d+/livecomment`（数字の識別子を示す `\d+` を使用）。
- リストBには、リストAにはない多くの追加パターンが含まれています。例えば `/api/user/me`、`/api/tag`、`/api/register` などです。
- いくつかのケースでは、パターンは似ていますが完全には同じではありません。例えば、リストAではライブストリームのパスに `\d+` を使用しています（数字の識別子を示す）、一方リストBでは `[^/]+` を使用しています（スラッシュ以外の任意の文字を示す、これには英数字も含まれる可能性があります）。
- リストBはより包括的で、より広範なAPIエンドポイントをカバーしています。

要約すると、二つのリストにはいくつかの重複がありますが、独自の要素も含まれており、類似したエンドポイントに対して少し異なるパターンを使用しています。

==============================

Reply:
[IMO] 差分はあれど、スクリプトの方を使うのが良いと思います🙋

Reply:
> 必ずしもe.GET/POST/...の形式ではないね👀
<任意の１文字以上の文字列>.GET/POST/...の形式に絞るのはどうだろう？

あれ、e消したつもりだったんですが残ってました、、、ありがとうございます！

https://github.com/isucon/isucon12-final/blob/d31d92c7f0df9fa1559270ffb9481ec8a610edf0/webapp/go/main.go#L86
おっと、初期実装でGroup使うパターンがあるんですね、見逃してました。`Group`使っていると
```go
API := e.Group("/hoge", h.apiMiddleware)
API.POST("/user", h.createUser)
```
みたいな時に検知失敗しちゃうんでちょっと考えないとですね、、、

Reply:
対応あざます👍
https://github.com/sugar-cat7/isucon-tool/pull/13/commits/abb35454f6e46abfd21542d71dba0c65b68e20f1#diff-9a5a3a7de8989a872be593a3b34fc94116082c443417b6e8d33c747101717bdeR10